src/java.base/share/classes/jdk/internal/foreign/abi/fallback/FFIType.java

```diff
@@ -127,17 +127,18 @@ private static void verifyStructType(GroupLayout grpl, List<MemoryLayout> filter
         try (Arena verifyArena = Arena.openConfined()) {
             MemorySegment offsetsOut = verifyArena.allocate(SIZE_T.byteSize() * filteredLayouts.size());
             LibFallback.getStructOffsets(structType, offsetsOut, abi);
-            for (int i = 0; i < filteredLayouts.size(); i++) {
-                MemoryLayout element = filteredLayouts.get(i);
-                final int finalI = i;
-                element.name().ifPresent(name -> {
-                    long layoutOffset = grpl.byteOffset(MemoryLayout.PathElement.groupElement(name));
-                    long ffiOffset = (long) VH_SIZE_T_ARRAY.get(offsetsOut, finalI);
-                    if (ffiOffset != layoutOffset) {
+            long expectedOffset = 0;
+            int offsetIdx = 0;
+            for (MemoryLayout element : grpl.memberLayouts()) {
+                if (!(element instanceof PaddingLayout)) {
+                    long ffiOffset = (long) VH_SIZE_T_ARRAY.get(offsetsOut, offsetIdx++);
+                    if (ffiOffset != expectedOffset) {
                         throw new IllegalArgumentException("Invalid group layout." +
-                                " Offset of '" + name + "': " + layoutOffset + " != " + ffiOffset);
+                                " Offset of '" + element.name().orElse("<unnamed>")
+                                + "': " + expectedOffset + " != " + ffiOffset);
                     }
-                });
+                }
+                expectedOffset += element.byteSize();
```

Comment:
I think the offset should only be added for structs, not for unions

Reply:
We can never encounter a union in this code. In libffi they are represented as a single element struct.

Reply:
Support for unions is not implemented in the fallback linker at the moment [1](https://github.com/openjdk/panama-foreign/pull/770/files#diff-b2d5d47f7117501ef2f8d7c497f0e2d999ac4374ae0a17aacaf12590cb2fc111R116). I wanted to do that as part of: https://bugs.openjdk.org/browse/JDK-8300294 when we have some tests, and there might be issues on other linkers as well that need fixing.

Reply:
I've changed the parameter type to `StructLayout` to make this more clear.